pkg/build/buildkit/connector/incluster.go

```diff
@@ -0,0 +1,226 @@
+			ic.ioCtrl.Logger().Infof("existing pod IP %s no longer valid: %s, will request new pod", ic.podIP, err)
+			ic.podIP = ""
+		} else {
+			ic.ioCtrl.Logger().Infof("verified buildkit connection is working, reusing pod IP: %s", ic.podIP)
+			return nil
+		}
+	}
+
+	// Request a new pod from the API
+	podIP, err := ic.assignBuildkitPod(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to assign buildkit pod: %w", err)
+	}
+	ic.podIP = podIP
+
+	return nil
+}
+
+// verifyConnection checks that we can connect to the current pod IP
+func (ic *InClusterConnector) verifyConnection(ctx context.Context) error {
+	// Use a short timeout for verification
+	verifyCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	c, err := ic.createBuildkitClient(verifyCtx)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	_, err = c.Info(verifyCtx)
+	return err
+}
+
+// assignBuildkitPod gets the least loaded buildkit pod with queue waiting support
+func (ic *InClusterConnector) assignBuildkitPod(ctx context.Context) (string, error) {
+	startTime := time.Now()
+	pollInterval := initialPollIntervalPortForward
+
+	sp := ic.ioCtrl.Out().Spinner("Waiting for BuildKit pod to become available...")
+	sp.Start()
+	defer sp.Stop()
+
+	for {
+		if time.Since(startTime) >= ic.maxWaitTime {
+			return "", fmt.Errorf("timeout waiting for buildkit pod after %v: please contact your cluster administrator to increase the maximum number of BuildKit instances or adjust the metrics thresholds", ic.maxWaitTime)
+		}
+
+		response, err := ic.oktetoClient.Buildkit().GetLeastLoadedBuildKitPod(ctx, ic.sessionID)
+		if err != nil {
+			return "", fmt.Errorf("could not get least loaded buildkit pod: %w", err)
+		}
+
+		if response.PodIP != "" {
+			ic.ioCtrl.Logger().Infof("assigned buildkit pod IP: %s", response.PodIP)
+			return response.PodIP, nil
+		}
+
+		if response.TotalInQueue > 0 {
+			friendlyReason := waitReasonMessages[response.Reason]
+			if friendlyReason == "" {
+				friendlyReason = response.Reason
+			}
+			ic.ioCtrl.Logger().Infof("Waiting for BuildKit: %s (position %d of %d in queue)",
+				response.Reason, response.QueuePosition, response.TotalInQueue)
+			sp.Stop()
+			sp = ic.ioCtrl.Out().Spinner(fmt.Sprintf("Waiting for BuildKit: %s (position %d of %d in queue)",
+				friendlyReason, response.QueuePosition, response.TotalInQueue))
+			sp.Start()
+		}
+
+		select {
+		case <-time.After(pollInterval):
+			pollInterval = time.Duration(float64(pollInterval) * backoffMultiplierPortForward)
+			if pollInterval > maxPollIntervalPortForward {
+				pollInterval = maxPollIntervalPortForward
+			}
+		case <-ctx.Done():
+			return "", fmt.Errorf("context cancelled while waiting for buildkit pod: %w", ctx.Err())
+		}
+	}
+}
+
+// createBuildkitClient creates a buildkit client with TLS certificates (like PortForwarder)
+func (ic *InClusterConnector) createBuildkitClient(ctx context.Context) (*client.Client, error) {
```

Comment:
Nit: as we talked offline, possible optimization for the future. Make a singleton client, and detect if the `ic.podIP` is the same that the one used for the existent client, and if don't, create a new one

Reply:
ok, sounds good!